common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -9977,9 +9976,14 @@ static jboolean NativeCrypto_usesBoringSsl_FIPS_mode() {
 
 static int NativeCrypto_BIO_read(JNIEnv* env, jclass, jlong bioRef, jbyteArray outputJavaBytes) {
     CHECK_ERROR_QUEUE_ON_RETURN;
-    BIO* bio = reinterpret_cast<BIO*>(static_cast<uintptr_t>(bioRef));
+    BIO* bio = to_BIO(env, bioRef);
     JNI_TRACE("BIO_read(%p, %p)", bio, outputJavaBytes);
 
+    if (bio == nullptr) {
```

Comment:
as above? Seems to_BIO is strict now.

Reply:
It is always strict but this is still possible (and necessary). "Throwing" a Java exception from JNI code doesn't change the flow of control, it just queues up the exception to be thrown on the next transition back to Java code so you still need to return early in the error  case.

It would be nice to combine the null check with an actual early return, which is my longer term plan for which this is just step 1.

One possibility is a macro that combines the call to `to_BIO()` with the null check and return (and JNI_TRACE), although most of these functions have different return types so it might get a little fugly. I _think_ recent C++ can do non-local returns from inline functions so that might be a more type-safe way to go.  But not this CL, if only because we're lacking unit tests for all the null checks in native code (which I have a WIP CL for, but it's a way off).

Reply:
Is that behaviour because of boringssl/google exception handling strategy? I thought you could throw from any point even across the boundary.

Reply:
Nope, it's the nature of JNI... https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#exception_handling

Reply:
Thanks, I suspect I learned some habits from e.g. https://github.com/facebookincubator/fbjni/blob/master/docs/rationale.md#exceptions

Reply:
Ah right.... Unfortunately Android [doesn't support](https://developer.android.com/training/articles/perf-jni#exceptions) C++ exceptions  so we can't unify them that way.